File: c++/bst.cpp
#include <bits/stdc++.h>
using namespace std;

struct Node
{
    int val;
    Node *left;
    Node *right;
};

struct Node *newNode(int val)
{
    struct Node *node = (struct Node *)malloc(sizeof(struct Node));

    node->val = val;
    node->left = NULL;
    node->right = NULL;

    return node;
}

void preOrder(struct Node *temp)
{
    if (temp != NULL)
    {
        cout << " " << temp->val;
        preOrder(temp->left);
        preOrder(temp->right);
    }
}

void inOrder(struct Node *temp)
{
    if (temp != NULL)
    {
        inOrder(temp->left);
        cout << " " << temp->val;
        inOrder(temp->right);
    }
}

void postOrder(struct Node *temp)
{
    if (temp != NULL)
    {
        postOrder(temp->left);
        postOrder(temp->right);
        cout << " " << temp->val;
    }
}

struct Node *insert(struct Node *node, int val)
{
    if (node == NULL)
    {
        return newNode(val);
    }
    if (val < node->val)
    {
        node->left = insert(node->left, val);
    }
    else
    {
        node->right = insert(node->right, val);
    }

    return node;
}

struct Node *minValueNode(struct Node *node)
{
    struct Node *temp = node;
    while (temp && temp->left != NULL)
    {
        temp = temp->left;
    }
    return temp;
}

struct Node *deleteNode(struct Node *root, int val)
{
    if (root == NULL)
        return root;

    if (val < root->val)
    {
        root->left = deleteNode(root->left, val);
    }
    else if (val > root->val)
    {
        root->right = deleteNode(root->right, val);
    }
    else
    {
        if (root->left == NULL)
        {
            struct Node *temp = root->right;
            free(root);
            return temp;
        }
        else if (root->right == NULL)
        {
            struct Node *temp = root->left;
            free(root);
            return temp;
        }
        struct Node *temp = minValueNode(root->right);
        root->val = temp->val;
        root->right = deleteNode(root->right, temp->val);
    }
    return root;
}

int main()
{
    struct Node *root = NULL;
    root = insert(root, 8);
    root = insert(root, 3);
    root = insert(root, 1);
    root = insert(root, 6);
    root = insert(root, 7);
    root = insert(root, 10);
    root = insert(root, 14);
    root = insert(root, 4);

    cout << "Inorder traversal: ";
    inOrder(root);

    cout << "\nAfter deleting 10\n";
    root = deleteNode(root, 10);
    cout << "Inorder traversal: ";
    inOrder(root);

    cout << "\nPreorder traversal: ";
    preOrder(root);
    cout << "\nPostorder traversal: ";
    postOrder(root);

    return 0;
}

File: pat/bst.py
class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

class bst:
    def __init__(self):
        self.root = None

    def insert_node(self, data):
        new_node = Node(data)
        if self.root is None:
            self.root = new_node
        else:
            curr_node = self.root
            while True:
                if data < self.root.data:
                    if curr_node.left is None:
                        curr_node.left = new_node
                        return
                    else:
                        curr_node = curr_node.left
                else:
                    if curr_node.right is None:
                        curr_node.right = new_node
                        return
                    else:
                        curr_node = curr_node.right

    def inorder_traversal(self, root): 
        print()
        if root is None:
            return
        self.inorder_traversal(root.left)
        print(root.data, end=" ")
        self.inorder_traversal(root.right)

    def preorder(self,root):
        print()
        if root is None:
            return
        print(root.data, end=" ")
        self.preorder(root.left)
        self.preorder(root.right)

    def postorder(self,root):
        print()
        if root is None:
            return
        self.postorder(root.left)
        self.postorder(root.right)
        print(root.data, end=" ")

tree = bst()
tree.insert_node(5)
tree.insert_node(10)
tree.insert_node(20)
tree.insert_node(4)
tree.insert_node(1)
tree.insert_node(19)
tree.insert_node(27)
tree.insert_node(45)
tree.insert_node(21)
tree.insert_node(14)
tree.insert_node(18)
tree.inorder_traversal(tree.root)
tree.preorder(tree.root)
tree.postorder(tree.root)